kotlinpoet/src/test/java/com/squareup/kotlinpoet/KmSpecsTest.kt

```diff
@@ -0,0 +1,326 @@
+  }
+
+  class ConstructorClass(val foo: String, vararg bar: Int) {
+    // Secondary constructors are ignored, so we expect this constructor to not be the one picked
+    // up in the test.
+    constructor(bar: Int) : this("defaultFoo")
+  }
+
+  @Test
+  fun supertype() {
+    val typeSpec = Supertype::class.toTypeSpec()
+
+    assertThat(typeSpec.superclass).isEqualTo(BaseType::class.asClassName())
+    assertThat(typeSpec.superinterfaces).containsKey(BaseInterface::class.asClassName())
+  }
+
+  abstract class BaseType
+  interface BaseInterface
+  class Supertype : BaseType(), BaseInterface
+
+  @Test
+  fun properties() {
+    val typeSpec = Properties::class.toTypeSpec()
+
+    assertThat(typeSpec.propertySpecs).hasSize(4)
+
+    val fooProp = typeSpec.propertySpecs.find { it.name == "foo" } ?: throw AssertionError(
+        "Missing foo property")
+    assertThat(fooProp.type).isEqualTo(String::class.asClassName())
+    assertThat(fooProp.mutable).isFalse()
+    val barProp = typeSpec.propertySpecs.find { it.name == "bar" } ?: throw AssertionError(
+        "Missing bar property")
+    assertThat(barProp.type).isEqualTo(String::class.asClassName().copy(nullable = true))
+    assertThat(barProp.mutable).isFalse()
+    val bazProp = typeSpec.propertySpecs.find { it.name == "baz" } ?: throw AssertionError(
+        "Missing baz property")
+    assertThat(bazProp.type).isEqualTo(Int::class.asClassName())
+    assertThat(bazProp.mutable).isTrue()
+    val listProp = typeSpec.propertySpecs.find { it.name == "aList" } ?: throw AssertionError(
+        "Missing baz property")
+    assertThat(listProp.type).isEqualTo(List::class.parameterizedBy(Int::class))
+    assertThat(listProp.mutable).isTrue()
+  }
+
+  class Properties {
+    val foo: String = ""
+    val bar: String? = null
+    var baz: Int = 0
+    var aList: List<Int> = emptyList()
+  }
+
+  @Test
+  fun companionObject() {
+    val typeSpec = CompanionObject::class.toTypeSpec()
+    assertThat(typeSpec.typeSpecs).hasSize(1)
+    val companionObject = typeSpec.typeSpecs.find { it.isCompanion }
+    checkNotNull(companionObject)
+    assertThat(companionObject.name).isEqualTo("Companion")
+  }
+
+  class CompanionObject {
+    companion object
+  }
+
+  @Test
+  fun namedCompanionObject() {
+    val typeSpec = NamedCompanionObject::class.toTypeSpec()
+    assertThat(typeSpec.typeSpecs).hasSize(1)
+    val companionObject = typeSpec.typeSpecs.find { it.isCompanion }
+    checkNotNull(companionObject)
+    assertThat(companionObject.name).isEqualTo("Named")
+  }
+
+  class NamedCompanionObject {
+    companion object Named
+  }
+
+  @Test
+  fun generics() {
+    val typeSpec = Generics::class.toTypeSpec()
+
+    assertThat(typeSpec.typeVariables).hasSize(3)
+    val tType = typeSpec.typeVariables[0]
+    assertThat(tType.name).isEqualTo("T")
+    assertThat(tType.isReified).isFalse()
+    assertThat(tType.variance).isNull() // we don't redeclare out variance
```

Comment:
Not clear what "redeclare" here means

Reply:
specifying `out` would be redundant

Reply:
Why so? Not sure I get it.

Reply:
if you write `List<out T>`, the compiler will nag you about it being unnecessary. This checks that we don't emit it that way